src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,563 @@
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(decimal? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(double value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(double? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(float value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(float? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(Guid value)
+            => value.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string? Format(Guid? value)
+            => value?.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string? Format(int value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(int? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(long value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(long? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private struct QueryStringBuilder
+        {
+            private readonly StringBuilder _builder;
+
+            private bool _hasNewParameters;
+
+            public string UriWithQueryString => _builder.ToString();
+
+            public QueryStringBuilder(ReadOnlySpan<char> uriWithoutQueryString)
+            {
+                _builder = new();
+                _builder.Append(uriWithoutQueryString);
+
+                _hasNewParameters = false;
+            }
+
+            public void AppendParameter(ReadOnlySpan<char> encodedName, ReadOnlySpan<char> encodedValue)
+            {
+                if (!_hasNewParameters)
+                {
+                    _hasNewParameters = true;
+                    _builder.Append('?');
+                }
+                else
+                {
+                    _builder.Append('&');
+                }
+
+                _builder.Append(encodedName);
+                _builder.Append('=');
+                _builder.Append(encodedValue);
+            }
+        }
+
+        private class ParameterData
+        {
+            public string? EncodedValue { get; private set; }
+            public bool DidReplace { get; set; }
+
+            public ParameterData(string? encodedValue)
+            {
+                EncodedValue = encodedValue;
+                DidReplace = false;
+            }
+        }
+
+        private struct EncodedParameterNameEqualityComparer : IEqualityComparer<ReadOnlyMemory<char>>
+        {
+            public bool Equals(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y)
+                => x.Span.SequenceEqual(y.Span);
```

Comment:
Should this be case insensitive instead? Like here:
https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Components/Components/src/Routing/QueryParameterValueSupplier.cs#L191-L202

Reply:
`SequenceEqual` is already case-insensitive, but I've changed this for the sake of consistency 😃

Reply:
I realize now I said case-insensitive. It was case sensitive and I misinterpreted your feedback. Apologies!